Issue 7069

melton-jason opened the pull request:
Fixes #7069 

Caused by the changes in #6581. 

Specifically, 

> @specify/dev-testing 
> I have modified the frontend Collection API such that if a dependent collection is initialized with resources, it makes sure to first create the empty collection and then explicitly add the resources to the collection (triggering any onAdd events). 
> Essentially, in terms of frontend business rules all `customInit` behavior specific for dependent collections can be moved/merged to `onAdd` : which is hopefully more intuitive. What do you all think? 

The `onAdded` business rule of `Determination` for example did not account for existing Determinations on existing CollectionObjects: 

https://github.com/specify/specify7/blob/4c3ccc496622d614b35d7129c953111dae5b1150/specifyweb/frontend/js_src/lib/components/DataModel/businessRuleDefs.ts#L417-L437

The BackBone ORM would receive the CollectionObject from the backend, create the empty Determination Collection, and then iteratively add Determinations in the order they appeared in the response (which triggered the onAdded sequentially for each Determination). 

### Checklist

- [ ] Self-review the PR after opening it to make sure the changes look good and
      self-explanatory (or properly documented)
- [ ] Add relevant issue to release milestone
- [ ] Add pr to documentation list
- [ ] Add automated tests
- [ ] Add a reverse migration if a migration is present in the PR

### Testing instructions
(Taken from #6581, with added tests) 

For Determination: 
- In App Resource -> Remote Preferences, make sure the following preference is present: CO_CREATE_DET_<collection_id>=true
  - For more information, see the Speciforum: [Automatically Creating a new Preparation, Determination, or Collection Object Attribute upon Collection Object creation](https://discourse.specifysoftware.org/t/automatically-create-a-new-preparation-determination-or-collection-object-attribute-upon-co-creation/1014) 
- Navigate to a new Collection Object form in Data Entry
- [ ] Ensure a Determination is being automatically created and that the `Is Current` checkbox is true
- [ ] Add a new Determination to the Collection Object and ensure the new Determination is now current
- [ ] Ensure the old Determination now has the `Is Current` checkbox unchecked
- [ ] Check the `Is Current` checkbox on the first Determination and ensure no other Determinations have the `Is Current` checkbox unchecked

- Open an existing CollectionObject with more than one Determination. 
- [ ] Ensure the first Determination is the current Determination
- Make an edit to the CollectionObject and save
- [ ] Ensure the Determination that was previously current is still current
- Add a new Determination to the CollectionObject and save
- [ ] Ensure the lastly added Determination is the current Determination
- [ ] Reload the page (if needed) and ensure the aforementioned Determination is displayed as the fist Determination
- Make a change to the CollectionObject and save
- [ ] Ensure that the previous Determination is still current 

- (If needed) Add `isPrimary` to the Collector form, and `collectors` to the Collecting Event form
- Navigate to a new Collecting Event form (or Collection Object form with Collecting Event as subview) 
- [ ] Add a Collector and ensure the Collector is the primary Collector
- [ ] Add another Collector and ensure the first Collector is still the primary Collector
- [ ] Check the second Collector as primary and ensure the previous Collector no longer has the `is Primary` checkbox checked
- [ ] Uncheck the second Collector's `Is Primary` checkbox and ensure the first Collector is no the primary Collector

lexiclevenger reviewed (commented):
### Testing instructions
(Taken from #6581, with added tests) 
- [ ] Ensure a Determination is being automatically created and that the `Is Current` checkbox is true
- [x] Add a new Determination to the Collection Object and ensure the new Determination is now current
- [x] Ensure the old Determination now has the `Is Current` checkbox unchecked
- [x] Check the `Is Current` checkbox on the first Determination and ensure no other Determinations have the `Is Current` checkbox unchecked

- Open an existing CollectionObject with more than one Determination. 
- [x] Ensure the first Determination is the current Determination
- Make an edit to the CollectionObject and save
- [x] Ensure the Determination that was previously current is still current
- Add a new Determination to the CollectionObject and save
- [x] Ensure the lastly added Determination is the current Determination
- [x] Reload the page (if needed) and ensure the aforementioned Determination is displayed as the fist Determination
- Make a change to the CollectionObject and save
- [x] Ensure that the previous Determination is still current 

- (If needed) Add `isPrimary` to the Collector form, and `collectors` to the Collecting Event form
- Navigate to a new Collecting Event form (or Collection Object form with Collecting Event as subview) 
- [x] Add a Collector and ensure the Collector is the primary Collector
- [x] Add another Collector and ensure the first Collector is still the primary Collector
- [x] Check the second Collector as primary and ensure the previous Collector no longer has the `is Primary` checkbox checked
- [x] Uncheck the second Collector's `Is Primary` checkbox and ensure the first Collector is no the primary Collector

- On an existing CollectingEvent with Collectors
- [ ] Ensure `isPrimary` behavior is expected adn identical with behavior in `v7.10.2.3`

---

The auto-created determination does not have isCurrent checked, and the user can save without it being checked. 


https://github.com/user-attachments/assets/a6dd2b62-3179-4d14-a673-cef287115a76

https://kufish20250214-issue-7069.test.specifysystems.org/specify/

The collector behavior matches the instructions, but the behavior seems broken in 7.10.2.3 in every case I've tried, so I can't really compare them. It definitely works much better in this branch, though!

issue-7069:

https://github.com/user-attachments/assets/4b9f9c15-2cdd-4051-a4fd-525b8d7424bd



7.10.2.3:

https://github.com/user-attachments/assets/356b4cbd-bc1a-4581-9953-0a584361200e


https://kufish20250214-v71023.test.specifysystems.org/specify/view/collectingevent/13809/

grantfitzsimmons reviewed (changes requested):
I had the same issue as Lexi– automatically created determinations are unfortunately not set as current.

---
- In App Resource -> Remote Preferences, make sure the following preference is present: CO_CREATE_DET_<collection_id>=true
  - For more information, see the Speciforum: [Automatically Creating a new Preparation, Determination, or Collection Object Attribute upon Collection Object creation](https://discourse.specifysoftware.org/t/automatically-create-a-new-preparation-determination-or-collection-object-attribute-upon-co-creation/1014) 
- Navigate to a new Collection Object form in Data Entry
- [ ] Ensure a Determination is being automatically created and that the `Is Current` checkbox is true
- [X] Add a new Determination to the Collection Object and ensure the new Determination is now current
- [X] Ensure the old Determination now has the `Is Current` checkbox unchecked
- [X] Check the `Is Current` checkbox on the first Determination and ensure no other Determinations have the `Is Current` checkbox unchecked

- Open an existing CollectionObject with more than one Determination. 
- [X] Ensure the first Determination is the current Determination
- Make an edit to the CollectionObject and save
- [X] Ensure the Determination that was previously current is still current
- Add a new Determination to the CollectionObject and save
- [X] Ensure the lastly added Determination is the current Determination
- [X] Reload the page (if needed) and ensure the aforementioned Determination is displayed as the fist Determination
- Make a change to the CollectionObject and save
- [X] Ensure that the previous Determination is still current 

- (If needed) Add `isPrimary` to the Collector form, and `collectors` to the Collecting Event form
- Navigate to a new Collecting Event form (or Collection Object form with Collecting Event as subview) 
- [X] Add a Collector and ensure the Collector is the primary Collector
- [X] Add another Collector and ensure the first Collector is still the primary Collector
- [X] Check the second Collector as primary and ensure the previous Collector no longer has the `is Primary` checkbox checked
- [X] Uncheck the second Collector's `Is Primary` checkbox and ensure the first Collector is no the primary Collector

- On an existing CollectingEvent with Collectors
- [X] Ensure `isPrimary` behavior is expected adn identical with behavior in `v7.10.2.3`

melton-jason commented:
Thank you for the reviews @grantfitzsimmons and @lexiclevenger! 
I forgot I improved on the `Collector -> isPrimary` behavior in #6581, so the behavior should be different (and hopefully better than) `v7.10.2.3`! I've updated the testing instructions. 

I also fixed the bug with automatically created determinations having their `isCurrent` set.

grantfitzsimmons reviewed (approved):
- [X] Ensure a Determination is being automatically created and that the `Is Current` checkbox is true
- [X] Add a new Determination to the Collection Object and ensure the new Determination is now current
- [X] Ensure the old Determination now has the `Is Current` checkbox unchecked
- [X] Check the `Is Current` checkbox on the first Determination and ensure no other Determinations have the `Is Current` checkbox unchecked
- [X] Ensure the first Determination is the current Determination
- Make an edit to the CollectionObject and save
- [X] Ensure the Determination that was previously current is still current
- Add a new Determination to the CollectionObject and save
- [X] Ensure the lastly added Determination is the current Determination
- [X] Reload the page (if needed) and ensure the aforementioned Determination is displayed as the fist Determination
- Make a change to the CollectionObject and save
- [X] Ensure that the previous Determination is still current 
- [X] Add a Collector and ensure the Collector is the primary Collector
- [X] Add another Collector and ensure the first Collector is still the primary Collector
- [X] Check the second Collector as primary and ensure the previous Collector no longer has the `is Primary` checkbox checked
- [X] Uncheck the second Collector's `Is Primary` checkbox and ensure the first Collector is no the primary Collector
- [X] Ensure the correct Collector is marked as Primary
- Add a new Collector to the CollectingEvent and ensure the prior Collector is still the Primary Collector 
- [X] Check the second Collector as primary and ensure the previous Collector no longer has the `is Primary` checkbox checked
- [X] Uncheck the second Collector's `Is Primary` checkbox and ensure the first Collector is no the primary Collector

specifysoftware commented:
This pull request has been mentioned on **Specify Community Forum**. There might be relevant details there:

https://discourse.specifysoftware.org/t/data-entry-bug-in-7-11-0-fix-pending/2708/1

melton-jason commented:
Thanks for the tests everyone! This branch was used as the base for the `v7.11.0.2` release: https://github.com/specify/specify7/commits/v7.11.0.2/. 

For the PR which brings these changes to `main`, see #7078
